public/app/features/explore/state/actions.ts

```diff
@@ -416,111 +424,210 @@ export function modifyQueries(
+  // but we're using the datasource interval limit for now
+  const minInterval = datasourceInstance.interval;
+  const datasourceName = exploreItemState.requestedDatasourceName;
+  const timeZone = getTimeZone(getState().user);
+
+  if (appendQueries && datasourceId !== 'loki') {
+    // currently append queries work only with Loki due to lack of universal line limit settings
+    // does not affect other types of queries
+    return;
+  }
 
-    // Some datasource's query builders allow per-query interval limits,
-    // but we're using the datasource interval limit for now
-    const minInterval = datasourceInstance.interval;
+  if (!hasNonEmptyQuery(queries)) {
+    dispatch(clearQueriesAction({ exploreId }));
+    dispatch(stateSave()); // Remember to save to state and update location
+    return;
+  }
 
-    stopQueryState(querySubscription);
+  // filtering out hidden queries
+  const nonHiddenQueries: DataQuery[] = queries.filter((query: DataQuery) => !query.hide);
 
-    const datasourceId = datasourceInstance.meta.id;
+  // maps queries in append queries mode to get only half of results
+  const updatedQueries: LokiQuery[] = appendQueries
+    ? nonHiddenQueries.map((query: LokiQuery) => {
+        const datasource = (datasourceInstance as unknown) as LokiDatasource;
+        const limit = query.maxLines ? query.maxLines : datasource.maxLines;
+        const maxLines = limit ? Math.floor(limit / 2) : 500;
+        return { ...query, maxLines };
+      })
+    : [];
+
+  const allResponseTimestamps: string[] = queryResponse.series
+    .map((elem: any) => {
+      return elem.fields[0].values.toArray();
+    })
+    .flat()
+    .sort((a: string, b: string) => a.localeCompare(b));
+
+  /**
+   * When using appendQueries we have to add 1s offset, with milisecond-level offsets it fetches the same log lines again
+   * can result in skipped log lines when linesLimit or maxLines params are set to the number
+   * smaller than the number of log lines during the given second.
+   */
+  const oldestResponseEndTimestamp = toUtc(allResponseTimestamps[0]).subtract('1', 's');
+
+  let appendQueryRaw: RawTimeRange | undefined = undefined; // updated raw time range for append queries, undefined for other types of queries
+  let appendQueryTimeRange: TimeRange | undefined = undefined; // updated time range for append queries, undefined for other types of queries
+
+  if (appendQueries) {
+    if (direction === QueryDirection.backward) {
+      appendQueryRaw = {
+        from: processDateTime(range.raw.from),
+        to: processDateTime(oldestResponseEndTimestamp),
+      };
+    } else if (prevShowMoreLogsTimeRanges.length < 2) {
+      appendQueryRaw = {
+        from: processDateTime(range.raw.from),
+        to: processDateTime(range.raw.to),
+      };
+    } else {
+      appendQueryRaw = {
+        from: processDateTime(prevShowMoreLogsTimeRanges[prevShowMoreLogsTimeRanges.length - 2].raw.from),
```

Comment:
Why do we want the one before last here not sure I understand the case here.

Reply:
ah ok so the length - 2 range is basically the next time range in the future in context of the logs timeline. This is not very clear here tbh. Especially with the `prevShowMoreLogsTimeRanges.length < 2` branch it looks like here are 3 cases on the same level but this basically just does forward/backward and for forward there is a special case for `prevShowMoreLogsTimeRanges.length < 2` case.

Reply:
Good catch. I'll try to rewrite it to make it more readable and less confusing. And yes, `prevShowMoreLogsTimeRanges.length - 2` is just the next one in the future context of the logs timeline.